### PR TITLE
Improve WebSocket communication

### LINKED ISSUE #50: Handle same user connection from different location
Suppose a user is connected to a session S from client L and attempts to connect to S on a different client X (a different device or different tab)

*Current Behavior*
- The server sends an error to client L stating that the user has already been connected to session S and refuses connection from client X.

*Proposed Behavior*
- The server should accept the connection from client X and close the connection from client L with a message to client L saying the user has been connected at a different location.

Discord has the same behavior as the proposed behavior above and I believe this would be better behavior than the current behavior at the moment. Rather than receiving an error for joining on the new client, the user can start playing the game instead on the new client.

### PR DESCRIPTION
**Pull request by Ramko9999**
1. Enabled the same client to connect from different connections while ensuring old connections are closed #50. 
2. Added modal to alert the client of WebSocket disconnection (comment below displaying the modal) #51.
3. Batched multiple messages directed to the same target #52 .
4. Replaced `useSocket` with `useGameConnection` to handle WS logic on the client.
5. Displayed the repository and file name above the editor for the current source code (comment below displaying "before" and "after").

### CODE REVIEW

**Comment by Ramko9999** [comment]
2. The picture of the disconnection modal:
image

**Comment by Ramko9999** [comment]
5. The picture of the repo and file name displayed above the editor:

Before:
image

After:
image

File: api-server/routes/ws.py [hunk 1]
```diff
@@ -308,10 +343,18 @@ async def on_delete_comment(delete_comment_body: DeleteCommentBody):
...
             except KeyError as e:
                 LOGGER.warn(
                     f"Invalid request body: 'message_type' key missing from request"
                 )
-    except WebSocketDisconnect:
+    except WebSocketDisconnect as e:
+        LOGGER.info(f"Client disconnection: {e.code} {e.reason}")
+        if e.code != WSAppStatusCodes.SWITCHING_CONNECTIONS:
+            await on_leave()
+    except WSAppPolicyViolation as e:
```
Comment (Ramko9999) [inline api-server/routes/ws.py:355]: Calling `connection.close` inside `on_pick`, `on_add_comment` and etc... did not seem to raise a `WebSocketDisconnect`, but rather a `RuntimeError` complaining that data cannot be received from a closed WebSocket [here](https://github.com/rohanshiva/gitgame/pull/54/files#diff-86e706b25d7746fe55a6a7acbdc477711f2b95ff09538eb5b7e4eab35dda7e3dR335). 

Suppose I have 2 WebSockets: A, B. Interestingly, if I called `A.close` from B's event loop (I speculate, although with little confidence, that the app processes each WebSocket in a separate event loop), `WebSocketDisconnect` gets raised in A's event loop. I need to investigate a bit further to understand the reason why the exception is raised in the latter scenario but not the former. 

Due to the above inconvenience, I simply raised a `WSAppPolicyViolation` within `on_pick`, `on_add_comment` and etc... and handled closing the connection and leaving in the catch handler.

File: client/src/components/game/hooks/gameConnection/UseGameConnection.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        const batchPayload = payload as BatchPayload;
+        const alertPayload = batchPayload.messages.find(
+          ({ message_type }) => message_type === ResponseType.ALERT
+        ) as AlertPayload;
+        if (alertPayload) {
+          onAlert(alertPayload.alert);
+        }
+      }
+
+      // AlertPayloads are ignored in the reducer
```
Comment (Ramko9999) [inline client/src/components/game/hooks/gameConnection/UseGameConnection.tsx:53]: I want to restrict the `GameReducer` to only performing state manipulations. `AlertPayload` messages only trigger toasts to display,  not alter the game state. This is why I ignored `AlertPayload` messages in the `GameReducer` and handled them directly above.

**Comment by Ramko9999** [comment]
There may still be a few bugs with the WS backend logic. I will setup testing for that later.

File: api-server/models.py [hunk 3]
```diff
@@ -36,6 +35,7 @@ class NoSelectedSourceCodeError(Exception):
 class Session(models.Model):
     id = fields.CharField(max_length=20, pk=True)
     created_at = fields.DatetimeField(auto_now_add=True)
+    version = fields.IntField(default=0)
```
Comment (rohanshiva) [inline api-server/models.py:38]: what is version?

File: client/src/components/game/hooks/gameConnection/UseGameConnection.tsx (hunk 2 above)
Comment (rohanshiva) [reply client/src/components/game/hooks/gameConnection/UseGameConnection.tsx:53]: Makes sense.

**Comment by rohanshiva** [comment]
LGTM, we can probably delete `Playground` stuff.

File: api-server/models.py (hunk 3 above)
Comment (Ramko9999) [reply api-server/models.py:38]: I kept it there to debug any "concurrent" updation issues.

**Pull request merged**